src/main/java/org/apache/maven/plugins/pmd/exec/CpdReportConsumer.java

```diff
@@ -0,0 +1,119 @@
+import org.codehaus.plexus.util.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class CpdReportConsumer implements Consumer<CPDReport> {
+    private static final Logger LOG = LoggerFactory.getLogger(CpdReportConsumer.class);
+
+    private final CpdRequest request;
+    private final ExcludeDuplicationsFromFile excludeDuplicationsFromFile;
+
+    CpdReportConsumer(CpdRequest request, ExcludeDuplicationsFromFile excludeDuplicationsFromFile) {
+        this.request = request;
+        this.excludeDuplicationsFromFile = excludeDuplicationsFromFile;
+    }
+
+    @Override
+    public void accept(CPDReport report) {
+        try {
+            // always create XML format. we need to output it even if the file list is empty or we have no
+            // duplications so that the "check" goals can check for violations
+            writeXmlReport(report);
+
+            // html format is handled by maven site report, xml format has already been rendered
+            // a renderer is only needed for other formats
+            String format = request.getFormat();
+            if (!"html".equals(format) && !"xml".equals(format)) {
+                writeFormattedReport(report);
+            }
+        } catch (MavenReportException e) {
+            // Exceptions happening during this consumer are logged by PMD
+            // through the CPDConfiguration's reporter, which is in this case PmdErrorMessageReporter.
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void writeXmlReport(CPDReport cpd) throws MavenReportException {
+        File targetFile = writeReport(cpd, new XMLRenderer(request.getOutputEncoding()), "xml");
+        if (request.isIncludeXmlInSite()) {
+            File siteDir = new File(request.getReportOutputDirectory());
+            siteDir.mkdirs();
+            try {
+                FileUtils.copyFile(targetFile, new File(siteDir, "cpd.xml"));
+            } catch (IOException e) {
+                throw new MavenReportException("Error while copying cpd.xml report to site dir", e);
+            }
+        }
+    }
+
+    private void writeFormattedReport(CPDReport cpd) throws MavenReportException {
+        CPDReportRenderer r = CpdExecutor.createRenderer(request.getFormat(), request.getOutputEncoding());
+        writeReport(cpd, r, request.getFormat());
+    }
+
+    private File writeReport(CPDReport cpd, CPDReportRenderer r, String extension) throws MavenReportException {
+        if (r == null) {
+            return null;
+        }
+
+        File targetDir = new File(request.getTargetDirectory());
+        targetDir.mkdirs();
+        File targetFile = new File(targetDir, "cpd." + extension);
+        try (Writer writer = new OutputStreamWriter(new FileOutputStream(targetFile), request.getOutputEncoding())) {
+            r.render(cpd.filterMatches(filterMatches()), writer);
+            writer.flush();
+        } catch (IOException ioe) {
```

Comment:
I'm tempted to let this one bubble up and convert it higher up the stack if needed